types/hapi__joi/index.d.ts

```diff
@@ -2054,6 +2054,11 @@ declare namespace Joi {
          */
         in(ref: string, options?: ReferenceOptions): Reference;
 
+        /**
+         * Checks whether or not the provided argument is an instance of ValidationError
+         */
+        isError(error: any): boolean;
```

Comment:
Can you make `ValidationError` a proper class as it is exported as one and make `isError` return `error is ValidationError`?

Reply:
Updating `isError` to return `error is ValidationError` is easy and makes sense, but I'm not sure what all would go into making `ValidationError` "a proper class". Whether it's a class or an interface depends on implementation, not definition, right? And since the official `@hapi/joi` library just adds some properties to the base `Error` object, I don't know how defining it any differently than it is will make a difference.

Then again, I'm not a TypeScript guru, so maybe there's something I'm not understanding.

Reply:
@SimonSchick I looked into it and learned that there is a reason to declare `ValidationError` as a class: it enables you to say `new Joi.ValidationError()` within TypeScript without throwing an error. I also read the `@hapi/joi` source to verify that `Joi.ValidationError` ***was*** a class before continuing...

However after defining `Joi.ValidationError` as a class, a ton of other things broke in the definition. Apparently at the bottom of `hapi__joi/index.d.ts` there was a line re-declaring `Joi` in order to export `Joi.Root` instead of the whole `Joi` namespace:

```
declare namespace Joi {
    // ...
    interface Root {
        // ...
    }
    // ...
}

declare const Joi: Joi.Root;
export = Joi;
```

I don't know if this is a standard practice in TypeScript definitions (possibly to keep enums and other definitions hidden within the namespace and avoid polluting a global namespace), but the moment I changed `interface ValidationError extends Error {` to `class ValidationError extends Error {` this duplicate identify `Joi` started throwing errors (the duplicate identifier didn't throw an error when `ValidationError` was an interface)

Not sure how to proceed from here -- need someone with a better understanding of TypeScript to chime in.

Reply:
Mhhh, ok that's a weird side effect, I will look into that.

Reply:
Yea I can't get it to work either, the problem is that the `Joi` namespace only holds types UNTIL you add the class keyword when then makes it conflict with the declaration merge.

For now let's just go with `isError(error: any): error is ValidationError` and defer the class export to some other time.

Reply:
This morning I was derailed on another project, but before I make that change I wanted to read through the TypeScript docs and give this one more good attempt. See if I can understand how declaration files work, what's up with the `export = Joi` syntax (which doesn't look like valid JS to me), and see if I can understand what they were attempting to do with the `Joi.Root` interface and if I can refactor this to fix it

If I can't get it working in a couple of hours I'll give up and go with `isError(error: any): error is ValidationError`

Reply:
@SimonSchick so I'm about ready to give up on this. I read through the TypeScript docs and they weren't helpful -- 99% of the docs are for basic things like how to declare a function return type. Even among the basic things it was clear that the `@hapi/joi` definitions aren't following best practices, because the TypeScript docs said your definition files should have the same directory structure as your library, meaning we should have several files like `errors.d.ts`, `messages.d.ts`, `validator.d.ts`, etc.

I was going to just update the definition with `isError()` when I noticed something in the block comment at the top:

```
// Type definitions for @hapi/joi 16.0
```

I then checked the current version: https://hapi.dev/module/joi/

The latest is 17.1.1

Is there a proper process for updating this version number to avoid conflicts with anyone using DefinitelyTyped in the wild? If someone pulls in the `@types/hapi__joi` package and they're on v16 I don't want any changes I make that are v17-only to mess with them.